Fix #3009: Question project backend final milestone

giritheja opened the pull request:
This PR completes #3009 .
**Checklist**
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

giritheja committed:
milestone 5 implementation

giritheja committed:
changes

giritheja committed:
lint issues

giritheja committed:
lint issues

giritheja commented:
@seanlip There is still some work to be done to fix delete method in question test.

apb7 commented:
Hello @giritheja! I think there are still some linting issues (I ran lint checks on your code locally). Can you please update your fork and run `python scripts/pre_commit_linter.py --path=.` once to confirm the same, or better so, run `bash scripts/run_presubmit_checks.sh`?  
Thanks!

giritheja committed:
Merge remote-tracking branch 'upstream/develop' into questions_p3

* upstream/develop: (77 commits)
  Add test to HintAndSolutionManagerService to check that hints aren't released after a correct answer is submitted. (#4585)
  Chance cancel property in SpeechSynthesisChunkerService into its own variable. (#4591)
  Enable indent for func-expr (#4588)
  Update text of refresher exp modal. (#4584)
  Update content-type for JSON responses. (#4583)
  Prevent hints from being released after correct answer is submitted. (#4580)
  Change fraction placeholder when no integer part is allowed. (#4581)
  Fix #4567: Changed condition to check range enclosed criteria, edited existing test (#4575)
  Fix part of #3905: Added semi-rule (#4576)
  Adds proper message for validation error in skill ID. (#4577)
  Fix part of #3905: Add checks for correct docstring style. (#4572)
  Fix part of #4374: Add docstrings to core.domain.collection_domain. (#4569)
  Add check to prevent 'SpeechSynthesisUtterance is undefined' error. (#4571)
  Fix of #2533: Modified the stripFormatting filter to preserve newlines in RTE cut-and-paste (#4568)
  Fix part of #3905: Add check for console.log statement (#4564)
  Change Java prerequisite to v8 (#4570)
  Fix #1542: RTE Components correctly displayed in answer group header (#4565)
  Fix part of #4374: Add docstrings to core.domain.param_domain (#4551)
  Fix #4555: Address review comments for #4536. (#4556)
  Update changelog, AUTHORS, CONTRIBUTORS and credits. (#4560)
  ...

giritheja committed:
update delete method

giritheja committed:
lint issues.

giritheja committed:
lint issues.

giritheja commented:
Hey @apb7 , Thank you.
@seanlip I have one small issue I am not sure about. After clearing that issue, I think this PR looks good for review.

giritheja commented on core/controllers/question_test.py:
```diff
@@ -111,3 +111,216 @@ def test_get(self):
+
+        response = self.testapp.delete(
+            '%s/' % feconf.QUESTION_DATA_URL, expect_errors=True)
+        self.assertEqual(response.status_int, 405)
```
I am not sure how to test the case where collection id and question id are not provided in this setting. I tried %s -> feconf.QUESTION_DATA_URL, which throws a 500 for some reason. :/

seanlip replied:
500 usually means "look at the stack trace of the error". I think you'll need to find a way to view the stack trace, since requests to handlers should never return 500 in general.

The simplest way I can think of to do it would be to add a call from the frontend of a random page (say, the About page or the library) that calls the given endpoint with an $http['delete'](...), then start up a dev server and look at the server logs to see what the 500 error stack trace looks like.

giritheja replied:
Thanks @seanlip . Fixed it.

giritheja committed:
fix delete issue

giritheja committed:
lint issue

giritheja committed:
lint issue

pranavsid98 commented:
Hi @seanlip, I think this PR needs a review. Thanks!

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+            raise self.PageNotFoundException
+        question = question_domain.Question.from_dict(
+            self.payload.get('question'))
+        question_id = question_services.add_question(
```
Hm. So how do questions get added to collections/skills?

giritheja replied:
I think the idea is to add question first and then update the question with skill id, which then adds question to skill.  I think passing skill_id along with the question in payload and adding question to skill as soon as it is made will be a better idea.

seanlip replied:
OK, sounds good. I agree that questions should always be attached to skills.

giritheja replied:
updated.

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+            raise self.PageNotFoundException
+        if not self.payload.get('change_list'):
+            raise self.PageNotFoundException
+        change_list = [question_domain.QuestionChange(
```
```
change_list = [
    question_domain.QuestionChange(change)
    for change in json.loads(self.payload.get('change_list'))]
```

Btw why do you need the json.loads()? Isn't it already unjsonified?

giritheja replied:
change_list in payload is list of dicts which jsonified and need to be unjsonified and converted to QuestionChange objects.

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+        change_list = [question_domain.QuestionChange(
+            change) for change in (
+                json.loads(self.payload.get('change_list')))]
+        question_services.update_question(
```
```
        question_services.update_question(
            self.user_id, collection_id, question_id, change_list,
            commit_message)
```

giritheja replied:
updated

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+    @acl_decorators.open_access
+    def delete(self, collection_id=None, question_id=None):
+        """Handles Delete requests."""
+        # print self.request.get('collection_id')
```
Why all these lines?

giritheja replied:
updated.

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+        # print self.request.get('collection_id')
+        # collection_id = self.payload.get('collection_id')
+        # question_id = self.payload.get('question_id')
+        print collection_id, question_id
```
Drop print statements.

giritheja replied:
updated

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+        })
+
+    @acl_decorators.open_access
+    def delete(self, collection_id=None, question_id=None):
```
Why are there default args here? Can we get rid of them? If the other URLs don't take params then this shouldn't too.

giritheja replied:
This is an issue. In test_utils, I found get_json, put_json and post_json but did not find delete_json. So, I had to use testapp's delete method. The above three methods accept payload but not delete method in testapp. I tried printing payload which always resulted in variable not found.
It was surprising to note that delete is different from other URLs. I have checked other controllers, The above approach was utilized to overcome this issue. So, I have employed the same.
I think creating a separate issue to develop delete method in test_utils should help. What do you think?

seanlip replied:
Hm, can we do `def delete(self, collection_id, question_id):` (without the None)?

giritheja replied:
We can do that but in the long run this will be reoccurring issue.

seanlip replied:
Let's do it without the None for this PR, and then create a separate issue to implement delete_json in test_utils as you suggest.

giritheja replied:
updated.

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,77 @@ def get(self):
+        collection_id = self.request.get('collection_id')
+        if not collection_id:
+            raise self.PageNotFoundException
+        question_summaries = (
```
```
        question_summaries = ([
            question_summary.to_dict() for question_summary in
            question_services.get_question_summaries_for_collection(
                collection_id)])

seanlip replied:
Except that the name of the variable is wrong. Shouldn't it be question_summary_dicts?

Also, I recommend doing, instead:

    question_summaries = question_services.get_question_summaries_for_collection(collection_id)

and then below:

    return self.render_json({
        'question_summary_dicts': [question_summary.to_dict() ...]
    })

giritheja replied:
updated.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -111,3 +111,216 @@ def test_get(self):
+        self.assertIn('question_id', response_json.keys())
+
+        del self.payload['question']
+        response = self.testapp.post(
```
You're reinventing the wheel. Look up self.post_json, self.put_json, self.get_json in our codebase. Use those instead, throughout.

giritheja replied:
I know it's weird but this is a fix to make tests pass.
The tests with 200 will only pass if we have self.post_json, self.put_json ... and 404 cases will only pass with testapp. A switch will cause all the tests to fail. I need to dwell deeper to figure out why this is happening.

seanlip replied:
I thought there were expect_errors, expected_status_int args that you can use for the error cases?

giritheja replied:
Yes, I am using them for error cases. I think I am not clear on my issue.
Here is a better explanation:
When I am expecting 404 as status_int for the above line,
I get 404 as status_int when I access the link with testapp but when I access the link using put_json, I get 200 as status_int, which is not equal to 404 and the test fails.

seanlip replied:
That seems odd, and actually unexpected -- I believe we use put_json successfully elsewhere to detect error cases, see e.g. `test_email_functionality_cannot_be_used_by_non_moderators`. I think this needs more investigation.

Did you also use the expect_errors arg when you tested? Were all environmental conditions the same? If you can explain how to reproduce the error (e.g. by swapping out a self.testapp.put for a put_json) then I would be happy to look into it.

giritheja replied:
I tried the following code
<img width="683" alt="screen shot 2018-02-23 at 12 37 58 pm" src="https://user-images.githubusercontent.com/8845039/36616093-685b729e-1897-11e8-9ae5-dd383d2397c4.png">
It gives me this:
<img width="600" alt="screen shot 2018-02-23 at 12 38 34 pm" src="https://user-images.githubusercontent.com/8845039/36616117-72e3489a-1897-11e8-96e5-03a72478d3b6.png">
I tried this:
<img width="604" alt="screen shot 2018-02-23 at 1 08 52 am" src="https://user-images.githubusercontent.com/8845039/36616134-82ba700e-1897-11e8-9e62-ea43b8f4c0e7.png">
I get this:
https://pastebin.com/rgr487gX

seanlip replied:
Not sure why you're passing the dict with a 'payload' key in the second clause. What if you follow the first one and just pass self.payload instead?

giritheja replied:
The dict was used to overcome one of the testapp checks. I have tried removing dict for self.put_json, I get status_int as 200 instead of 404 even though PageNotFoundException is raised.

seanlip replied:
I think you are reading the error message wrongly. If you do not pass in an expected_status_int then the test infrastructure assumes you mean a 200 response. Please make sure to read put_json carefully and look at other examples of its use in the codebase to understand how it works.

I tried the following and it passes:

```
        del self.payload['change_list']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)

        del self.payload['commit_message']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)

        del self.payload['collection_id']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)

        del self.payload['question_id']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)
```

Also I think there is an error with the way this test is written. At each stage you should take the original payload and delete one field from it. Otherwise all the tests might be failing just because you deleted the first change_list field, causing subsequent checks to be irrelevant.

giritheja replied:
Yep, that fixes put_json but when I tried this fix on get_json, I still had issues.
```        
            response = self.get_json(
            '%s/batch' % feconf.QUESTION_DATA_URL,
            expect_errors=True, expected_status_int=404)

        del self.payload['stringified_skill_ids']
        response = self.get_json(
            '%s/batch' % feconf.QUESTION_DATA_URL, self.payload,
            expect_errors=True, expected_status_int=404)
```
I got
``` TypeError: get_json() got an unexpected keyword argument 'expected_status_int' ```
Should I use testapp here?

seanlip replied:
Nope. I suggest looking at the code for get_json() and its other uses in the codebase -- there ought to be other cases which result in 404 errors.

seanlip replied:
(That's how I figured out how to handle put_json, btw.)

giritheja replied:
resolved this for put and post requests but for get requests, it has to do something with the way the response are generated.
I get the following error:
`  File "/Users/giritheja/Desktop/opensource/oppia/core/tests/test_utils.py", line 274, in _parse_json_response
    json_response.content_type, 'application/json')
AssertionError: 'text/html' != 'application/json'`

seanlip commented on core/domain/question_domain.py:
```diff
@@ -236,3 +236,36 @@ def update_question_data(self, question_data):
+    Attributes:
+        question_id: str. The ID of the question.
+        question_title: str. The title of the question.
+        skill_ids: list(str). The list of skill IDs associted with question.
```
... **associated** with **the** question.

giritheja replied:
updated.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -236,3 +236,36 @@ def update_question_data(self, question_data):
+        Args:
+            question_id: str. The ID of the question.
+            question_title: str. The title of the question.
+            skills: list(str). The list of skill names associted with question.
```
associated

Also, docstring is not the same as line 247 which says skill_ids. Which should it be? **Please make sure to proofread your PR.**

giritheja replied:
Updated.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -236,3 +236,36 @@ def update_question_data(self, question_data):
+        self.skills = skills
+
+    def to_dict(self):
+        """Returns a dictionary of Question Summary domain object.
```
Returns a dictionary representation of this domain object.

giritheja replied:
updated.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -236,3 +236,36 @@ def update_question_data(self, question_data):
+        """Returns a dictionary of Question Summary domain object.
+
+        Returns:
+            Dict. A dict of Question Summary object.
```
dict. A dict representing this QuestionSummary object.

giritheja replied:
updated.

seanlip commented on core/domain/question_domain_test.py:
```diff
@@ -162,3 +162,17 @@ def test_update_methods(self):
+    """Test for Question Summary Domain object."""
+
+    def test_to_dict(self):
+        expected_object = {
```
This is not an object. It's a dict. Call it expected_object_dict.

giritheja replied:
updated.

seanlip commented on core/domain/question_services.py:
```diff
@@ -69,18 +69,28 @@ def add_question(committer_id, question):
             one.
+
+    Raises:
+        Exception. The question with ID is not present in collection with ID.
```
... not present in the given collection.

giritheja replied:
I am sorry this is little confusing, we are only passing collection id, not collection itself.
Should it be okay to mention collection?
More specifically, is it okay to be vague about the parameters in docs but not vague while naming the variables?

seanlip replied:
That's a fair question. I'm OK with saying `... not present in the collection with the given collection ID`, but I think a dev will understand the meaning of both.

giritheja replied:
updated.

seanlip commented on core/domain/question_services.py:
```diff
@@ -69,18 +69,28 @@ def add_question(committer_id, question):
+    question = get_question_by_id(question_id)
+    if collection_id != question.collection_id:
+        raise Exception(
+            'The question with ID %s is not present in collection with ID %s' % ( # pylint: disable=line-too-long
```
No, just break in the middle of the string.

```
    raise Exception(
        '.....'
        '.....' % (..., ...))
```

giritheja replied:
updated.

seanlip commented on core/domain/question_services.py:
```diff
@@ -218,21 +228,32 @@ def _save_question(committer_id, question, change_list, commit_message):
             question.
+
+    Raises:
+        Exception. The question with ID is not present in collection with ID.
```
... in the given collection.

giritheja replied:
updated.

seanlip commented on core/domain/question_services.py:
```diff
@@ -218,21 +228,32 @@ def _save_question(committer_id, question, change_list, commit_message):
+    question = get_question_by_id(question_id)
+    if collection_id != question.collection_id:
+        raise Exception(
+            'The question with ID %s is not present in collection with ID %s' % ( # pylint: disable=line-too-long
```
Same note as above. Do what you can to avoid needing to use pragmas where possible.

giritheja replied:
updated.

seanlip commented on core/domain/question_services.py:
```diff
@@ -304,3 +325,30 @@ def get_questions_batch(
+        list(QuestionSummary). A list of Question Summary objects.
+    """
+    collection = collection_services.get_collection_by_id(collection_id)
+    skill_ids_of_question = {}
```
This is confusing, it suggests that only one question is being talked about. Perhaps call it questions_to_skill_ids?

Also consider using collections.defaultdict(list). Then you might not need lines 343-344.

giritheja replied:
updated.

seanlip commented on core/domain/question_services.py:
```diff
@@ -304,3 +325,30 @@ def get_questions_batch(
+        for question_id in skill.question_ids:
+            if question_id not in skill_ids_of_question:
+                skill_ids_of_question[question_id] = []
+            skill_ids_of_question[question_id].append(skill.name)
```
skill.name is not the same as a skill ID. Be careful with the naming and make sure you're adding the correct thing to the dict.

giritheja replied:
It is skill name. Updated.

seanlip commented on core/domain/question_services_test.py:
```diff
@@ -99,7 +99,13 @@ def test_delete_question(self):
         question_id = question_services.add_question(self.owner_id, question)
-        question_services.delete_question(self.owner_id, question_id)
+        with self.assertRaisesRegexp(Exception, (
+            'The question with ID %s is not present in collection with ID random' % # pylint: disable=line-too-long
```
Ditto -- avoid pragmas. See above.

Please fix all future occurrences below, too.

giritheja replied:
Updated.

seanlip commented on core/domain/question_services_test.py:
```diff
@@ -99,7 +99,13 @@ def test_delete_question(self):
+        with self.assertRaisesRegexp(Exception, (
+            'The question with ID %s is not present in collection with ID random' % # pylint: disable=line-too-long
+            (question_id))):
+            question_services.delete_question(self.owner_id, 'random', (
```
It is cleaner to read this if you break after `delete_question(`. Ditto below in multiple places. That way, you also avoid introducing an unnecessary set of parens.

giritheja replied:
updated.

seanlip commented on feconf.py:
```diff
@@ -579,6 +579,7 @@ def get_empty_ratings():
 NEW_EXPLORATION_URL = '/contributehandler/create_new'
 PREFERENCES_DATA_URL = '/preferenceshandler/data'
 QUESTION_DATA_URL = '/questionhandler'
+QUESTION_MANAGE_URL = '/questionmanager'
```
QUESTION_MANAGER_URL = '/questionmanagerhandler'

giritheja replied:
updated.

seanlip commented on main.py:
```diff
@@ -468,6 +468,15 @@ def ui_access_wrapper(self, *args, **kwargs):
+        question.QuestionsHandler),
+    get_redirect_route(
+        r'%s/<collection_id>/<question_id>' % feconf.QUESTION_DATA_URL,
+        question.QuestionsHandler),
```
Nope -- one handler per URL please.

giritheja replied:
This is consequence of incapability of delete method to take in payload.
In order to keep uniformity, I will have to change put, post, get methods. Lemme know if you want me to do that.

seanlip replied:
Ah, I see. I'd follow how we do it in editor.py (ExplorationHandler): let's have one URL and handler for GET/PUT/DELETE that has collection_id and question_id as parameters. Then let's have a separately-named URL and handler for POST. Would that work?

giritheja replied:
Yeah, that works.

giritheja replied:
One issue with that, PUT takes in parameters question_id, collection_id, commit message and change_list whereas DELETE takes in collection_id and question_id. If we were to have a common handler for PUT and DELETE while deriving parameters from URL then for the PUT we need to pass the parameters commit message and change_list through payload, which doesn't seem right to me. What do you think?
Also, There is not GET for QuestionHandler.

seanlip replied:
Passing commit message and change list through payload is fine. We do a similar thing in the ExplorationHandler. It keeps the URLs short. Again, I suggest taking a look at ExplorationHandler to see how it's done there, and mimicking it.

Re GET, ah ok, then we don't need that method then.

giritheja replied:
updated.

seanlip reviewed (changes requested):
Hi @giritheja, sorry for the delayed review on this. I think this PR is pretty good overall in terms of structure, but the details needs some work.

Also, in addition to addressing the comments, could you please write a backend integration test (similar to [this](https://github.com/oppia/oppia/blob/develop/core/controllers/editor_test.py#L1099) and others) that does the following:
- creates a question and adds it to a collection
- creates a second question
- retrieves the list of question summaries
- deletes a question
- retrieves the list of question summaries

(or similar?) This backend integration test should only use get_json, post_json, put_json, delete_json and should not modify internal stuff. The aim is to make sure that the backend functionality works well as a black box when everything is chained together -- that way, you'll know that what you've built actually works end-to-end.

Other than that, it looks good! Please let me know if you have any questions about any of the comments, and ping me when you'd like a second look. Thanks!

giritheja committed:
addressed review comments

giritheja committed:
Merge remote-tracking branch 'upstream/develop' into questions_p3

* upstream/develop: (114 commits)
  Fix line too long (#4798)
  Fix #4374: Docstrings added to core/domain/stats_jobs_one_off.py (#4757)
  Fixes #4055: Removes redundant css selector from conversation_skin_directive.html (#4790)
  Fix #3971: add activity indicator to feedback updates (#4785)
  Fix part of #4374 : Update docstrings to core.domain.stats_domain. (#4731)
  Fix #2620 (#4791)
  Removes unused angular dependecies injected into controllers (#4792)
  Fix #4367 : Prevent null-submission for MathExpressionInput interaction (Upgraded guppy 260da6--->b5055b) (#4753)
  Upgrade GAE version to 1.9.67. (#4786)
  Fix #4781: Fixed zoom in Problem with code editor (#4783)
  Changed timeout for popup to 4 seconds. (#4706)
  Fixes #3982 : Added 'open' in the open feedback thread summary. (#4767)
  Fix #4778: Changed is_community_owned to make_community_owned. (#4780)
  Revert "Fixes #4530: Removes spaces and new lines from end of string in RTE. (#4748)" (#4782)
  Fixed #4537: Navigating between admin pages on url change (#4777)
  Added $log dependency in missing places. (#4772)
  Add missing dependency. (#4769)
  Fix #3478: Added functionality to sort nodes list. (#4708)
  Fixes #4530: Removes spaces and new lines from end of string in RTE. (#4748)
  Fixes #4410: Fixed info modal for private explorations (#4754)
  ...

giritheja committed:
lint issues

giritheja committed:
lint issues

giritheja committed:
lint issues

giritheja commented:
Hello @seanlip, sorry for the delay. I was very busy with academics. I have addressed review comments and added integration tests. PTAL.

anmolshkl commented on core/jobs.py:
```diff
@@ -33,8 +33,8 @@
 from mapreduce import mapreduce_pipeline
-from mapreduce import model as mapreduce_model
 from mapreduce import output_writers
+from mapreduce import model as mapreduce_model
```
@giritheja the import order seems to be incorrect. Could you please fix it? Thanks.

giritheja replied:
<img width="586" alt="screen shot 2018-03-15 at 11 33 20 am" src="https://user-images.githubusercontent.com/8845039/37484067-b9311a54-2844-11e8-82b0-c8d786084eeb.png">
I get the exact opposite error when I try to make changes and push. :/

seanlip replied:
@giritheja Try merging the latest version of upstream/develop into this branch? I think @apb7 made changes here a little while back.

(@apb7, any thoughts?)

giritheja replied:
<img width="759" alt="screen shot 2018-03-15 at 10 30 44 pm" src="https://user-images.githubusercontent.com/8845039/37505179-a3f19748-28a0-11e8-8224-982d2fda76e7.png">
Merged and trying to push but still facing same issue.

seanlip replied:
Hm, I'll have to defer you to @apb7 then. @apb7, could you help please?

(Also, @giritheja, thank you for checking!)

seanlip replied:
@apb7 -- ping?

apb7 replied:
Sorry for the delay, @seanlip! Looking into this.

apb7 replied:
Hi @giritheja, we recently changed the import-order in this file. Did you follow the steps as stated in this [discussion](https://groups.google.com/forum/#!topic/oppia-dev/mVzwx0TDip0)?  
The steps are as follows: 
1. Run `bash scripts/clean.sh`
2. Run `bash scripts/start.sh`
3. Run `python scripts/pre_commit_linter.py --path=./core/jobs.py`

Since you have already pulled the latest version of `develop` branch, isort should not report any errors now. This would solve the issue.  
(@seanlip: If this does not help, I will look into this further!)
Thanks!

seanlip replied:
I just tried this (for another PR) and it works -- thanks @apb7. It did result in an additional lint error in statistics/gae_models but I've fixed that (see https://github.com/oppia/oppia/pull/4818/commits/b7af371fba65c00d7be4331c38a98f0298239d05).

apb7 replied:
Thank you @seanlip! Re [`b7af371`](https://github.com/oppia/oppia/commit/b7af371fba65c00d7be4331c38a98f0298239d05), I think this is the only way to deal with it :)

anmolshkl reviewed (changes requested):
@giritheja the build seems to be breaking due to minor lint issues. PTAL!

giritheja committed:
lint issue

giritheja committed:
Merge remote-tracking branch 'upstream/develop' into questions_p3

* upstream/develop:
  Fix part of #3905: Add check for period at comment and docstring end (#4759)
  Remove numpy as a dependency. (#4808)
  Fix part of #3905: Add check for keyword arguments (#4752)
  Fix #4056:There is no way to add an exploration/collection into "play-later" list for mobile users. (#4709)
  Fixes part of #4374: Add docstring to user_query_services.py & user_services.py (#4796)
  Fix #4776: possible race condition in audio preview in editor tab. (#4802)
  Fix #4773: Add responsive behavior to the landing page (#4801)
  Fix part of #4364 : Enhancements to the correctness footer (#4770)

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,82 @@ def get(self):
+    """This handler completes PUT/DELETE requests for questions."""
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    @acl_decorators.open_access
```
Why is this open access? Shouldn't it require login at the very least?

Ditto elsewhere. I think, for now, restrict these handlers to admin/moderator level since the system is still in beta. Safer, and prevents changes to the questions datastore by random users before it's ready for full use. Add a backend test to ensure that users with permissions below this level can't use these handlers.

giritheja replied:
updated.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -111,3 +111,350 @@ def test_get(self):
+                'new_value': [self.skill_id]
+            }], 'Update skill')
+
+        self.payload = {}
```
Defining a self.payload in setUp() doesn't really add anything. Just have a 'payload' var in each test, or construct the dict manually inline. The less stuff you put on "self", the better -- it just adds to the amount of context that needs to be carried around.

Ditto for other test classes.

giritheja replied:
Updated.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -111,3 +111,350 @@ def test_get(self):
         self.assertEqual(response.status_int, 404)
+
+
+class QuestionsHandlerTest(test_utils.GenericTestBase):
```
All the docstrings for the newly added test classes are "Test the question handler" or variants thereof. This does not help to distinguish the test classes. Could you please rename the test classes to indicate which parts they're testing and update the docstrings to describe clearly the difference between the different classes?

giritheja replied:
Updated.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -111,3 +111,350 @@ def test_get(self):
+class QuestionsPostHandlerTest(test_utils.GenericTestBase):
+    """Test the question handler."""
+
+    def setUp(self):
```
If the setUp is the same as the above class then why not just combine the two classes? You can have more than one test method in a class. For each test method, setUp() will be run before that test, and all test methods are run independently.

Ditto below. In general try not to duplicate lots of code.

giritheja replied:
Updated.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -111,3 +111,350 @@ def test_get(self):
+        self.payload = {}
+
+
+    def test_black_box(self):
```
No clue what this does from the name. Please explain more clearly, either using a docstring or inline comments or a better function name or something.

giritheja replied:
updated.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -236,3 +236,38 @@ def update_question_data(self, question_data):
+    Attributes:
+        question_id: str. The ID of the question.
+        question_title: str. The title of the question.
+        skills: list(str). The list of skill names associated with the
```
Shouldn't this be skill_names? In the backend, a Skill is an object that's different from a skill name.

giritheja replied:
updated.

seanlip commented on feconf.py:
```diff
@@ -569,6 +569,8 @@ def get_empty_ratings():
 PREFERENCES_DATA_URL = '/preferenceshandler/data'
 QUESTION_DATA_URL = '/questionhandler'
+QUESTION_MANAGER_URL = '/questionmanagerhandler'
+QUESTION_POST_URL = '/questionposthandler'
```
Let's call this QUESTION_CREATION_URL = '/questioncreationhandler'. We've never added "post" or "get" to the URL before.

giritheja replied:
updated.

seanlip reviewed (changes requested):
Hi @giritheja, it's looking good, thanks! Sorry for the delay in reviewing. Just a few more cleanups and this should be done.

giritheja committed:
addressed review comments

giritheja committed:
lint issue

giritheja committed:
lint issue

giritheja committed:
lint issues

giritheja committed:
Merge remote-tracking branch 'upstream/develop' into questions_p3

* upstream/develop:
  Fix #4714: Alignment of button and text in exploration setting  (#4811)
  Fix #4813: Icon position changed and inline css deleted (#4817)
  Reorder tabs. (#4809)
  Update AUTHORS and CHANGELOG for v2.6.2 (#4822)
  Routine update of translations (#4818)
  Fix #4663: Added CSS to fix overflowing content out of modal. (#4664)
  Show Issues Card on State Editor (#4744)
  Fixes part of Issue#4374 : Making changes. (#4795)

giritheja committed:
removed testapp

giritheja commented:
Hey @seanlip , I have addressed your review comments. One weird thing I noticed was that I have written 6 tests but when I run backend_tests.sh, I get only 5 tests were verified successfully. I tried printing, I observed that the first and last tests were run successfully. There was one test in the middle is which was being missed out. I tried re-linting but that did not work. I tried re-ordering the tests it worked, the test worked but resulted in other test replacing it miss out. Do you have any suggestions to deal with this?

seanlip commented on core/controllers/question_test.py:
```diff
@@ -362,63 +199,79 @@ def test_get(self):
+            '%s/%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id, question_id),
+            payload, expect_errors=True)
+        self.assertEqual(response.status_int, 302)
```
This is weird -- 302 is not an error. Shouldn't it just be expected_status_int=302?

giritheja replied:
updated in 8ec2d3e.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -228,128 +97,96 @@ def test_delete(self):
+        self.assertEqual(response.status_int, 404)
 
-        question_id = question_services.add_question(self.owner_id, question)
+    def manager_get(self):
```
You need to prefix all test methods with `test_`. Otherwise they won't run.

seanlip commented on core/controllers/question_test.py:
```diff
@@ -228,128 +97,96 @@ def test_delete(self):
-        self.skill_id = collection.get_skill_id_from_skill_name('test')
-        collection_node = collection.get_node(self.exp_id)
-        collection_node.update_acquired_skill_ids([self.skill_id])
+        response = self.testapp.get(
```
Here and elsewhere -- I thought we talked about using self.get_json() instead of self.testapp.get()?

Please check the rest of the PR and fix accordingly.

giritheja replied:
I think I have mentioned the reason for using self.testapp.get in the previous comments.
So, whenever I try using get_json for the failing conditions, I get the following error:
`Traceback (most recent call last):
  File "/Users/giritheja/Desktop/opensource/oppia/core/controllers/question_test.py", line 164, in test_batch_get
    expect_errors=True)
  File "/Users/giritheja/Desktop/opensource/oppia/core/tests/test_utils.py", line 408, in get_json
    json_response, expect_errors=expect_errors)
  File "/Users/giritheja/Desktop/opensource/oppia/core/tests/test_utils.py", line 398, in _parse_json_response
    json_response.content_type, 'application/json')
AssertionError: 'text/html' != 'application/json'
`
I am not sure how to sort this out, so I used testapp.get

seanlip commented on core/controllers/question_test.py:
```diff
@@ -362,63 +199,79 @@ def test_get(self):
-        self.payload['question'] = question.to_dict()
-        self.payload['skill_id'] = self.skill_id
+        self.logout()
+        response = self.testapp.put(
```
self.put_json

giritheja replied:
updated in 8ec2d3e.

seanlip reviewed (commented):
Thanks @giritheja! Just some notes on tests (including a response to your question).

giritheja committed:
addressed review changes

seanlip commented on core/controllers/question_test.py:
```diff
@@ -57,57 +63,256 @@ def setUp(self):
+
+        del payload['skill_id']
+        self.post_json(
+            '%s' % feconf.QUESTION_CREATION_URL, {'payload': (
```
Why do you still have a payload key here? This should follow exactly the way you did it in lines 115-117.

Ditto below.

giritheja replied:
Updated.

giritheja commented:
Hey @seanlip , I think you may have reviewed 1ad688d instead of 8ec2d3e . A couple of review comments were covered in the latest commit. I have addressed the remaining comments. PTAL

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,82 @@ def get(self):
+
+class QuestionsHandler(base.BaseHandler):
+    """This handler completes PUT/DELETE requests for questions."""
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
```
Why False? That seems wrong. The pattern should be similar to ExplorationPage and ExplorationHandler.

That might also address your get_json issues.

giritheja replied:
Updated.

seanlip reviewed (commented):
Hi @giritheja -- I'm afraid you'll need to try to debug and figure out what's going on with text/html vs application/json and fix that. This isn't a new pattern, and there are lots of controllers + tests that use get_json without issues; trying to revert to self.testapp.get() etc. is not the correct way to handle this. (It's likely that there is some other issue in the code and bypassing the test in this way just hides it but doesn't fix the error.)

I made a couple inline comments that might be related, but the upshot is that this PR still needs to avoid using self.testapp.get/post/put.

giritheja committed:
addressed review changes

giritheja committed:
address review comments

seanlip commented on core/controllers/question.py:
```diff
@@ -41,3 +42,80 @@ def get(self):
+            'question_summary_dicts': [
+                question_summary.to_dict()
+                for question_summary in question_summaries]
+            })
```
Deindent by 4 to match opening indent.

giritheja committed:
address review comments

anmolshkl reviewed (approved):
@giritheja lgtm!
